engine.go

```diff
@@ -143,101 +112,229 @@ func (e *Engine) AnalyzeQuery(
 ) (sql.Schema, sql.RowIter, error) {
-	return e.QueryWithBindings(ctx, query, nil)
+	return e.QueryNodeWithBindings(ctx, query, nil, bindings)
 }
 
-func (e *Engine) QueryWithBindings(
+// QueryNodeWithBindings executes the query given with the bindings provided. If parsed is non-nil, it will be used
+// instead of parsing the query from text.
+func (e *Engine) QueryNodeWithBindings(
 	ctx *sql.Context,
 	query string,
+	parsed sql.Node,
 	bindings map[string]sql.Expression,
 ) (sql.Schema, sql.RowIter, error) {
 	var (
-		parsed, analyzed sql.Node
-		iter             sql.RowIter
-		err              error
+		analyzed sql.Node
+		iter     sql.RowIter
+		err      error
 	)
 
-	finish := observeQuery(ctx, query)
-	defer finish(err)
+	if parsed == nil {
+		var err error
+		parsed, err = parse.Parse(ctx, query)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
 
-	parsed, err = parse.Parse(ctx, query)
+	err = e.authCheck(ctx, parsed)
 	if err != nil {
 		return nil, nil, err
 	}
 
-	var perm = auth.ReadPerm
-	var typ = sql.QueryProcess
-	switch parsed.(type) {
-	case *plan.CreateIndex:
-		typ = sql.CreateIndexProcess
-		perm = auth.ReadPerm | auth.WritePerm
-	case *plan.CreateForeignKey, *plan.CreateCheck, *plan.DropForeignKey, *plan.AlterIndex, *plan.CreateView,
-		*plan.DeleteFrom, *plan.DropIndex, *plan.DropView,
-		*plan.InsertInto, *plan.LockTables, *plan.UnlockTables,
-		*plan.Update:
-		perm = auth.ReadPerm | auth.WritePerm
+	analyzed, err = e.Analyzer.Analyze(ctx, parsed, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if len(bindings) > 0 {
+		analyzed, err = plan.ApplyBindings(analyzed, bindings)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	setQueriedDatabase(ctx, parsed)
+
+	schema, rowIter, err2 := e.beginTransaction(ctx, parsed)
+	if err2 != nil {
+		return schema, rowIter, err2
+	}
+
+	iter, err = analyzed.RowIter(ctx, nil)
+	if err != nil {
+		return nil, nil, err
 	}
 
+	autoCommit, err := isSessionAutocommit(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if autoCommit {
+		iter = transactionCommittingIter{iter}
+	}
+
+	return analyzed.Schema(), iter, nil
+}
+
+func (e *Engine) beginTransaction(ctx *sql.Context, parsed sql.Node) (sql.Schema, sql.RowIter, error) {
+	// Before we begin a transaction, we need to know if the database being operated on is not the one
+	// currently selected
+	transactionDatabase := determineTransactionDatabase(ctx, parsed)
+
+	// TODO: this won't work with transactions that cross database boundaries, we need to detect that and error out
+	beginNewTransaction := ctx.GetTransaction() == nil
+	if beginNewTransaction {
+		if len(transactionDatabase) > 0 {
+			database, err := e.Catalog.Database(transactionDatabase)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			tdb, ok := database.(sql.TransactionDatabase)
+			if ok {
+				tx, err := tdb.StartTransaction(ctx)
+				if err != nil {
+					return nil, nil, err
+				}
+				ctx.SetTransaction(tx)
+			}
+		}
+	}
```

Comment:
refactor suggestion: wrap the `transactionCommittingIter` here with the `transactionDatabase` embedded

Reply:
Good call, this revealed some other grossness